File: app/src/main/java/com/jmalinen/blackjack/model/Hand.kt
package com.jmalinen.blackjack.model

import com.jmalinen.blackjack.engine.HandEvaluator

data class Hand(
    val cards: List<Card> = emptyList(),
    val bet: Int = 0,
    val isSplitHand: Boolean = false,
    val isDoubledDown: Boolean = false,
    val isSurrendered: Boolean = false,
    val isStanding: Boolean = false,
    val splitFromAces: Boolean = false
) {
    val score: Int get() = HandEvaluator.bestScore(cards)

    val isSoft: Boolean get() = HandEvaluator.isSoft(cards)

    val isBusted: Boolean get() = score > 21

    val isBlackjack: Boolean
        get() = cards.size == 2 && score == 21 && !isSplitHand

    val isPair: Boolean
        get() = cards.size == 2 && cards[0].rank.baseValue == cards[1].rank.baseValue

    val isFinished: Boolean
        get() = isBusted || isStanding || isDoubledDown || isSurrendered || isBlackjack ||
                (splitFromAces && cards.size >= 2)

    fun addCard(card: Card): Hand = copy(cards = cards + card)
}

File: ios/Blackjack/Blackjack/Model/Hand.swift
struct Hand: Equatable {
    var cards: [Card] = []
    var bet: Int = 0
    var isSplitHand: Bool = false
    var isDoubledDown: Bool = false
    var isSurrendered: Bool = false
    var isStanding: Bool = false
    var splitFromAces: Bool = false

    var score: Int { HandEvaluator.bestScore(cards) }

    var isSoft: Bool { HandEvaluator.isSoft(cards) }

    var isBusted: Bool { score > 21 }

    var isBlackjack: Bool {
        cards.count == 2 && score == 21 && !isSplitHand
    }

    var isPair: Bool {
        cards.count == 2 && cards[0].rank.baseValue == cards[1].rank.baseValue
    }

    var isFinished: Bool {
        isBusted || isStanding || isDoubledDown || isSurrendered || isBlackjack ||
            (splitFromAces && cards.count >= 2)
    }

    func addCard(_ card: Card) -> Hand {
        var copy = self
        copy.cards = cards + [card]
        return copy
    }
}
